NoSpansCreatedException

zsistla opened the pull request:
See issue #219 

When calling tracer functions that deal with spans, `NoSpansCreatedException` is now thrown if no spans have been created.  Previously, `NoopSpans` were being silently returned, leaving the user no way of knowing whether a span was explicitly created or not.

Regarding adding `InvalidExporterEndpointException`, it seems like the exporters should handle their own endpoint validations, like the exception generated here for an invalid endpoint:
https://github.com/open-telemetry/opentelemetry-php/blob/main/contrib/Zipkin/Exporter.php#L63

It seems like it would be an exporter case by case basis depending what they need for their endpoints.

Added test cases.
Removed one test case (testActiveRootSpanParentIsNoopSpanIfNoParentProvided
) because it didn’t actually show what the code(https://github.com/open-telemetry/opentelemetry-php/blob/main/sdk/Trace/Tracer.php#L59) was doing where if no parent was passed in, the code sends in noop, but then a boilerplate context is generated.

morrisonlevi commented:
I don't fully understand what this is trying to solve. Can you explain more why a NoopSpan is problematic for end-users?

Also, throwing exceptions in telemetry code can be undesirable, because if left unhandled it will crash (or at least prematurely end). As much as possible, we don't want issues in telemetry code to affect the main app.

zsistla commented:
Hi @morrisonlevi ,
Thanks for feedback and good to see you pop in again :).

This was in response to #219 , but PHP Otel doesn't have a `dev mode` so the implementation went straight in with no dev mode toggle.

After addressing the concerns in issue #219 , I tend to agree with you.  I put the PR up to get discussion on it, but the silent Noop functionality seemed a lot cleaner.

Without a `dev mode` to filter the exception, throwing the exception seems to make it harder for the end user.  
I'd vote that unless we get implement/activate `dev mode`, the exception is less usable and we should table this PR and issue #219 for when we can get that mode.

@bobstrecansky what are you thoughts?

bobstrecansky commented:
My opinion is that a dev mode would be a very nice addition to this project.  I think we should probably open an issue about that and discuss how we want to execute that long term.

I do think the functionality in this PR would benefit more from a dev mode environment; a noop span does offer a lot of functionality.

Perhaps our next step is to discuss dev mode for this project and weave this into that process?

bobstrecansky commented:
We are planning on working on some core functionality before we start to implement a dev mode.  Closing for now.
